### PR TITLE
nixos/netbird-server: init module

### PR DESCRIPTION
**Pull request by Tom-Hubrecht**
## Description of changes

Adds a package for netbird dashboard and creates a module for self-hosting the netbird management server.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/networking/netbird-server.nix
Comment (RaitoBezarius) [inline nixos/modules/services/networking/netbird-server.nix:254]: Do you have to consistently recreate it or could something like rsync with appropriate flags could update the new files?

Comment (RaitoBezarius) [reply nixos/modules/services/networking/netbird-server.nix:254]: (explain why this is needed also)

File: nixos/modules/services/networking/netbird-server.nix
Comment (RaitoBezarius) [inline nixos/modules/services/networking/netbird-server.nix:250]: I suggest to do `StateDirectory = [ "netbird-mgmt" "netbird-mgmt/data" ]` which should work I believe

File: nixos/modules/services/networking/netbird-server.nix
Comment (RaitoBezarius) [inline nixos/modules/services/networking/netbird-server.nix:292]: forgotten

Comment (RaitoBezarius) [reply nixos/modules/services/networking/netbird-server.nix:292]: (and should probably be exposed somehow in the configuration)

File: nixos/modules/services/networking/netbird-server.nix
Comment (RaitoBezarius) [inline nixos/modules/services/networking/netbird-server.nix:322]: same comment on debug logging as above

File: pkgs/servers/web-apps/netbird-dashboard/default.nix
Comment (RaitoBezarius) [inline pkgs/servers/web-apps/netbird-dashboard/default.nix:7]: How close is the dashboard the original package? Should it be a passthru derivation of `pkgs.netbird` ?

i.e. `pkgs.netbird.dashboard` ?

File: nixos/modules/services/networking/netbird-server.nix
Comment (RaitoBezarius) [inline nixos/modules/services/networking/netbird-server.nix:212]: remove, not remove?

File: nixos/modules/services/networking/netbird-server.nix
Comment (RaitoBezarius) [inline nixos/modules/services/networking/netbird-server.nix:213]: upstream settings?

File: nixos/modules/services/networking/netbird-server.nix
Comment (RaitoBezarius) [inline nixos/modules/services/networking/netbird-server.nix:170]: why not just letting folks setting that directly?

File: nixos/modules/services/networking/netbird-server.nix
Comment (Tom-Hubrecht) [reply nixos/modules/services/networking/netbird-server.nix:254]: It is needed because some configuration values, e.g. the auth endpoint are hardcoded in the web-ui and replaced from env varibles with envsubst. I suppose that we could also pass those variables to the dashboard package and set them at build time.

File: nixos/modules/services/networking/netbird-server.nix
Comment (Tom-Hubrecht) [reply nixos/modules/services/networking/netbird-server.nix:212]: It is directly taken from https://github.com/netbirdio/netbird/blob/main/infrastructure_files/nginx.tmpl.conf, I guess we can remove it

File: nixos/modules/services/networking/netbird-server.nix
Comment (Tom-Hubrecht) [reply nixos/modules/services/networking/netbird-server.nix:213]: Same as above ( https://github.com/netbirdio/netbird/blob/main/infrastructure_files/nginx.tmpl.conf )

File: nixos/modules/services/networking/netbird-server.nix
Comment (Tom-Hubrecht) [reply nixos/modules/services/networking/netbird-server.nix:170]: Are you talking of the grant_type ?

File: nixos/modules/services/networking/netbird-server.nix
Comment (misuzu) [inline nixos/modules/services/networking/netbird-server.nix:31]: There should be a `meta.doc`, explaining how to actually use this module. There are a lot of options and it's not clear what a working "production"-like configuration should look like. A NixOS test is also desirable.

File: nixos/doc/manual/release-notes/rl-2311.section.md
Comment (misuzu) [inline nixos/doc/manual/release-notes/rl-2311.section.md:27]:
```suggestion
- A self-hosted management server for the [Netbird](https://netbird.io). Available as [services.netbird-server](#opt-services.netbird-server.enable).
```

File: nixos/modules/services/networking/netbird-server.nix
Comment (misuzu) [inline nixos/modules/services/networking/netbird-server.nix:107]:
```suggestion
    enableDeviceAuthorizationFlow = mkEnableOption "device authorization flow for netbird." // { default = true; };
```

File: nixos/modules/services/networking/netbird-server.nix
Comment (misuzu) [inline nixos/modules/services/networking/netbird-server.nix:116]: There should be a way to pass sensitive options via external file, all values passed to `services.netbird-server.settings` are going to be stored in nix store which is world-readable.

File: nixos/modules/services/networking/netbird-server.nix
Comment (misuzu) [inline nixos/modules/services/networking/netbird-server.nix:73]: There should be a link to the documentation describing these options.

File: nixos/modules/services/networking/netbird-server.nix
Comment (misuzu) [inline nixos/modules/services/networking/netbird-server.nix:216]: When reverse-proxy is used to handle ssl, `NETBIRD_DISABLE_LETSENCRYPT` should be enabled: https://docs.netbird.io/selfhosted/selfhosted-guide#configuration-for-netbird

Comment (Tom-Hubrecht) [reply nixos/modules/services/networking/netbird-server.nix:216]: Actually, it is only needed when you configure things with docker (see https://github.com/netbirdio/netbird/blob/main/infrastructure_files/configure.sh#L114 ), it is enough to pass no certificate or key file to the management config for it to work in our case

Comment (Tom-Hubrecht) [reply nixos/modules/services/networking/netbird-server.nix:216]: The way that netbird setups their docker image means that most env variables are not actually passed to the underlying images, and are just used when running `configure.sh`

File: nixos/modules/services/networking/netbird-server.nix
Comment (Tom-Hubrecht) [reply nixos/modules/services/networking/netbird-server.nix]: I added `secretFiles` options for this

File: nixos/modules/services/networking/netbird-server.nix
Comment (oddlama) [inline nixos/modules/services/networking/netbird-server.nix:289]: Can you guard this behind a new enable option? Some users might prefer to use a different reverse proxy (or proxy via another machine) so adding this unconditionally is probably not desirable.

File: nixos/modules/services/networking/netbird-server.nix
Comment (oddlama) [reply nixos/modules/services/networking/netbird-server.nix:346]: It's quite unfortunate that they require envsubst instead of accessing the environment variables directly... I don't like copying each time but it seems like this is currently necessary - unless upstream changes the process.

On that note: In https://github.com/netbirdio/dashboard/issues/242 one of the maintainers (?) noted that they would accept contributions to make this process easier to host without docker. I imagine that a simple PR to upstream that replaces this envsubst by just accessing `process.env.SomeVar` in their config.ts would solve this for us and for them.

**Review by oddlama — CHANGES_REQUESTED** [review]
Could you add a nixos tests that verifies that this is working as intended?

File: nixos/modules/services/networking/netbird-server.nix
Comment (Tom-Hubrecht) [reply nixos/modules/services/networking/netbird-server.nix:346]: I mean, we can also do it ourselves with substituteAll while building the dashboard, but we need to decide if we want to rebuild the dashboard as a whole every time an option change, which would take some time

Comment (oddlama) [reply nixos/modules/services/networking/netbird-server.nix:346]: IMO an upstream solution seems to be aligned with their goals anyway, so only if they refuse I would start resorting to alternatives.

File: nixos/modules/services/networking/netbird-server.nix
Comment (Tom-Hubrecht) [reply nixos/modules/services/networking/netbird-server.nix]: I have added a bit of documentation and also reduced the amount of settings that the end user has to provide by default. I don't have the time right now to write a complicated test, as I believe we need to setup multiple nodes, one being an sso and try adding nodes to the network in the test and until the start of september I will be a bit busy.

**Comment by mannp** [comment]
This will be very cool, once it is ready and merged, thank you 👍🏻

**Comment by RaitoBezarius** [comment]
@oddlama @misuzu do you have further comments on this PR or are you fine with it? It has been stuck for a while.

**Comment by misuzu** [comment]
> @oddlama @misuzu do you have further comments on this PR or are you fine with it? It has been stuck for a while.

Having an end-to-end NixOS test would be nice, though I'm not sure it'd be easy to implement. LGTM otherwise

**Comment by oddlama** [comment]
Generally this looks good to me, too. :)

I agree that tests would be great but maybe hard to implement properly. I've noticed some minor details that we can address later if required, for example it is always opening firewall ports right now. I do intend to give this a try soon, so if I notice anything else I'll probably make a small fixup PR.

This also currently modifies the 23.11 release notes, not sure whether that would be a problem for merging.

**Comment by Tom-Hubrecht** [comment]
I have to do a rebase, and put the release notes into 24.05, also update netbird-dashboard (2 version have passed since this PR)

**Comment by jvanbruegge** [comment]
I just stumbled upon this after being 90% done with setting this up for myself.
One thing that would be nice is being able to pass also the turn password via a file. This is how I did that:
```nix
services.coturn = mkIf cfg.createCoturn {
  enable = true;
  lt-cred-mech = true;
  no-cli = true;
  extraConfig = ''
    fingerprint
    no-software-attribute
    external-ip=${domain}
  '';
};

systemd.services.coturn.serviceConfig = {
  EnvironmentFile = "/run/secrets/netbird";
  ExecStart = lib.mkForce ''
    ${pkgs.coturn}/bin/turnserver -c /run/coturn/turnserver.cfg --user="''${COTURN_USER}:''${COTURN_PASSWORD}"
  '';
};
```
Also, the createNginx option could optionally accept a `listen` config, for people that use a seperate reverse proxy

File: nixos/modules/services/networking/netbird-server.nix
Comment (jvanbruegge) [inline nixos/modules/services/networking/netbird-server.nix:411]:
```suggestion
          cp -R ${cfg.dashboard}/* ${stateDir}/web-ui
```

File: nixos/modules/services/networking/netbird-server.nix
Comment (jvanbruegge) [inline nixos/modules/services/networking/netbird-server.nix:514]: There should be an additional option like `enableACME` to handle this. If used with a different reverse proxy this will not work

File: nixos/modules/services/networking/netbird-server.nix
Comment (jvanbruegge) [inline nixos/modules/services/networking/netbird-server.nix:377]: Additionally there could be a user defined secrets file here. Is a bit more convenient to have all the variables in one file

File: nixos/modules/services/networking/netbird-server.nix
Comment (jvanbruegge) [inline nixos/modules/services/networking/netbird-server.nix:502]: As I said earlier, it is possible to pass this via the command like. Also the config should probably set `external-ip` to `NETBIRD_DOMAIN`

File: nixos/modules/services/networking/netbird-server.nix
Comment (Tom-Hubrecht) [reply nixos/modules/services/networking/netbird-server.nix:514]: I mean, the option is named `configureNginx`, as I don't use any other reverse proxy, I don't feel confident maintaining them. The best course of action is to set `configureNginx` to `false` and configure your reverse proxy yourself.

File: nixos/modules/services/networking/netbird-server.nix
Comment (Tom-Hubrecht) [reply nixos/modules/services/networking/netbird-server.nix:502]: `external-ip` expects an ip address, which is not what `NETBIRD_DOMAIN` contains

Comment (jvanbruegge) [reply nixos/modules/services/networking/netbird-server.nix:598]: Still works as expected though. I tested this with coturn behind two layers of NAT

File: nixos/modules/services/networking/netbird-server.nix
Comment (jvanbruegge) [reply nixos/modules/services/networking/netbird-server.nix:623]: Yes and no. There should be a nginx config to host the dashboard.
Independent of this there could be a config to also reverse proxy the other two services incl acme.
This way, I could have nginx as plain webserver but still use a different reverse proxy

**Comment by jvanbruegge** [comment]
For comparison, I had to vendor this module and adapt it to fit my needs like this: https://github.com/jvanbruegge/server-config/blob/master/modules/netbird-server.nix

**Comment by PatrickDaG** [comment]
I took the module and reworked it a bit as well as updated everything to the newest versions:
- https://git.lel.lol/patrick/nix-config/src/branch/master/modules/netbird-server.nix
- https://git.lel.lol/patrick/nix-config/src/branch/master/modules/netbird-dashboard.nix

There are still a few thing missing in this draft most notably:
- systemd hardening
- secret management
- better documentation
But I wanted to post the draft here as a suggestion for further work.
# Notable things I've changed
## Split modules
as far as I see there is no reason that the dashboard and the server need to be on the same host.
I don't even think the different parts of the server, the management API and the signal server need to be on the same host
so to be completely flexible we might even want to to change those.
## dashboard templating
I've changed the dashboard templating to use a derivation so it's only done once at build time and not on every restart.
## Used RFC 42 style settings
Instead of this weird back and forth from Environment to templated json I've directly used json. I think this enables user to better configure the server to their need but it also forces us to always keep the defaults updated and might make configuring harder as the json values are pretty undocumented.
## OIDC autoconfig
 Turns out netbird is able to automatically get the oidc config from the AutoConfigEndpoint. It does so automatically and overwrites any values set in the config so we can just skip those.

# Things I'm still unsure about
## secret management
Some of those secrets are not a secret at all. Namely, the dashboard client secret is templated into the statically served frontend and thus open for anyone who can reach the webpage. For this reason I would oppose exposing this secret as a `file` option to not give user a false sense of security but make them very aware that if their Identity provider depends on this being secret that they should be careful.
Gladly, this secret is not needed most of the time, as OIDC supports `public clients` which should not need a secret.
As for the other client secret I'm not sure. I think netbird support using different IDP clients for different things, but my IDP does not support such a setup so I can't test if there is a need or use for the other secrets to be actually secret, in which case we should support them being read from files.

## IDP management
netbird has an IDP manager which can be used to read additional information from the IDP in a way that standart oidc does not support. But this only work if you use one of their supported IDPs, which I do not so again I'm unable to check if my config works for this case of if we need additional things.

## device Authorization
Same thing my IDP does no provide any device authorization so I can't test my config.

## Nginx setup
So far I've only exposed the nginx config for serving the frontend. You also may need additional config if you use nginx as a reverse proxy to correctly route all the other parts, management and signal. But as I don't want to assume people use nginx as a reverse proxy I think it would be better to just document what endpoints need to be reachable and let people set it up themselves.

**Comment by Tom-Hubrecht** [comment]
So, I rewrote the whole thing, thanks a lot for your work @PatrickDaG 

## Split modules

Each part is now activable independently : 
- signal service
- management service
- static dashboard
- coturn server

This split configuration also allows to setup nginx only for the parts that we want @jvanbruegge, the use of certificates for the coturn server is also independent now.

There is also the option to enable everything at once with `services.netbird.server.enable`

## Secrets

To use secrets in the management config, there is nifty function https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/utils.nix#L174 so we can either pass a value directly as a string that will end up in the store, or we can instead pass `{ _secret = "/pat/to/secret"; }` and the secret will be subtituted in the config at runtime.

File: nixos/modules/services/networking/netbird/management.nix
Comment (PatrickDaG) [inline nixos/modules/services/networking/netbird/management.nix:243]: I'm not sure if we should only listen on localhost, as anyone not using a reverse proxy on the same host cannot use it like this.

**Review by PatrickDaG — COMMENTED** [review]
Thanks for the work. Looks really good.
I've found some things I feel like we could improve. If you don't feel like it, I can also do it in the next week or so.

File: nixos/modules/services/networking/netbird/dashboard.nix
Comment (Tom-Hubrecht) [reply nixos/modules/services/networking/netbird/dashboard.nix]: Yes, it is available as `services.netbird.server.dashboard.finalDrv` now

File: nixos/modules/services/networking/netbird/coturn.nix
Comment (Tom-Hubrecht) [reply nixos/modules/services/networking/netbird/coturn.nix:137]: The turn rfcs say that by default there are alternative listening ports on the port + 1, I guess we should use the ports defined in the coturn config instead

File: nixos/modules/services/networking/netbird/management.nix
Comment (Tom-Hubrecht) [reply nixos/modules/services/networking/netbird/management.nix:243]: I mean, this is the default value, when using a more complex setup the user can always set a different one

**Comment by PatrickDaG** [comment]
@Tom-Hubrecht What's your status for this? Do you still have time and motivation to work on it? If not, I can offer to help out to get this merged.

**Comment by Tom-Hubrecht** [comment]
> @Tom-Hubrecht What's your status for this? Do you still have time and motivation to work on it? If not, I can offer to help out to get this merged.

I'll have time friday to address your comments, and then we should be good to go

File: nixos/modules/services/networking/netbird/coturn.nix
Comment (Tom-Hubrecht) [reply nixos/modules/services/networking/netbird/coturn.nix]: I did it with `LoadCredential` instead of changing the group, the change to the coturn module is the same as for the `passwordFile`, so it was straightforward

File: nixos/modules/services/networking/netbird/coturn.nix
Comment (Tom-Hubrecht) [reply nixos/modules/services/networking/netbird/coturn.nix]: I removed the `netbird.coturn.port` option, as it is just the `services.coturn.listening-port` one, and I used the ports defined in the `services.coturn` config instead for the firewall, with an option to customize which ones we want to open.

**Comment by PatrickDaG** [comment]
Looks great. Thanks for all the work.

File: nixos/modules/services/networking/netbird/coturn.nix
Comment (RaitoBezarius) [inline nixos/modules/services/networking/netbird/coturn.nix:41]:
```suggestion
        Whether to use ACME certificates corresponding to the given domain for the server.
```

File: nixos/modules/services/networking/netbird/coturn.nix
Comment (RaitoBezarius) [inline nixos/modules/services/networking/netbird/coturn.nix:145]: What if you made `coturn.service` a dependency on the `acme-something-something` ? Is there a reason not to do so?

Comment (RaitoBezarius) [reply nixos/modules/services/networking/netbird/coturn.nix:145]: (uphold or whatever is the right thing that systemd needs to be told)

File: pkgs/by-name/ne/netbird-dashboard/package.nix
Comment (RaitoBezarius) [inline pkgs/by-name/ne/netbird-dashboard/package.nix:22]: Can we have an issue for this patch to upstream and explain our needs?

**Review by RaitoBezarius — APPROVED** [review]
It's excellent work, almost idiomatic.

There's only one slight problem, it's the usage of the `enableNginx` pattern.
AFAIK the good way to do this is to provide a `nginx` subconfiguration fragment which reuse the NGINX types (look in other modules how they do it) and once you do something like `services.netbird.dashboard.nginx.enable = true;` or if the subfragment is non-null, you make use of NGINX.

This way, we can easily separate webservers implementation from the rest.

I do not deem this a blocker for this enormous PR and this is future work for anyone who care about other webservers and this ecosystem.

File: pkgs/by-name/ne/netbird-dashboard/package.nix
Comment (Tom-Hubrecht) [reply pkgs/by-name/ne/netbird-dashboard/package.nix:22]: I opened an issue upstream https://github.com/netbirdio/dashboard/issues/375

Comment (Tom-Hubrecht) [reply pkgs/by-name/ne/netbird-dashboard/package.nix]: Great news, netbird vendored the font in the latest release !

File: nixos/modules/services/networking/netbird/coturn.nix
Comment (Tom-Hubrecht) [reply nixos/modules/services/networking/netbird/coturn.nix:145]: As discussed in another channel, the fact that acme-thing.service is a (non-persistent) oneshot unit makes it quite complicated to do, one could use conflicts and onuccess to swap units but it is simpler for now to use `postRun`

**Comment by mannp** [comment]
Wow, thank you to all involved :) going to give this try 👍🏻

**Comment by srounce** [comment]
@Tom-Hubrecht I'm currently using this module and I'm curious why a Coturn submodule was added when a [Coturn module already existed](https://github.com/NixOS/nixpkgs/commits/master/nixos/modules/services/networking/coturn.nix) in NixOS? If I'm already using the other one should I point to that instead or should I run an instance from each module?

**Comment by Tom-Hubrecht** [comment]
> @Tom-Hubrecht I'm currently using this module and I'm curious why a Coturn submodule was added when a [Coturn module already existed](https://github.com/NixOS/nixpkgs/commits/master/nixos/modules/services/networking/coturn.nix) in NixOS? If I'm already using the other one should I point to that instead or should I run an instance from each module?

Becaus the current module does not allow specifying secrets so I had to create an overlay, under the hood it uses `services.coturn` so the options should merge

**Pull request merged**
